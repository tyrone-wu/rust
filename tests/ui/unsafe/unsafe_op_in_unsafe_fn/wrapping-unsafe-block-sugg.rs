//@ run-rustfix
//@ aux-build:external_unsafe_macro.rs

#![deny(unsafe_op_in_unsafe_fn)] //~ NOTE
#![crate_name = "wrapping_unsafe_block_sugg"]

extern crate external_unsafe_macro;

unsafe fn unsf() {}

pub unsafe fn foo() {
    //~^ NOTE an unsafe function restricts its caller, but its body is safe by default
    unsf(); //~ ERROR call to unsafe function `unsf` is unsafe
    //~^ NOTE call to unsafe function
    //~| NOTE for more information, see issue #71668
    //~| NOTE consult the function's documentation
    unsf(); //~ ERROR call to unsafe function `unsf` is unsafe
    //~^ NOTE call to unsafe function
    //~| NOTE for more information, see issue #71668
    //~| NOTE consult the function's documentation
}

pub unsafe fn bar(x: *const i32) -> i32 {
    //~^ NOTE an unsafe function restricts its caller, but its body is safe by default
    let y = *x; //~ ERROR dereference of raw pointer is unsafe and requires unsafe block
    //~^ NOTE dereference of raw pointer
    //~| NOTE for more information, see issue #71668
    //~| NOTE raw pointers may be null
    y + *x //~ ERROR dereference of raw pointer is unsafe and requires unsafe block
    //~^ NOTE dereference of raw pointer
    //~| NOTE for more information, see issue #71668
    //~| NOTE raw pointers may be null
}

static mut BAZ: i32 = 0;
pub unsafe fn baz() -> i32 {
    //~^ NOTE an unsafe function restricts its caller, but its body is safe by default
    let y = BAZ; //~ ERROR use of mutable static is unsafe and requires unsafe block
    //~^ NOTE use of mutable static
    //~| NOTE for more information, see issue #71668
    //~| NOTE mutable statics can be mutated by multiple threads
    let z = y + BAZ; //~ ERROR use of mutable static is unsafe and requires unsafe block
    //~^ NOTE use of mutable static
    //~| NOTE for more information, see issue #71668
    //~| NOTE mutable statics can be mutated by multiple threads
    z + bar(BAZ as *const _) //~ ERROR call to unsafe function `bar` is unsafe
    //~^ NOTE call to unsafe function
    //~| NOTE for more information, see issue #71668
    //~| NOTE consult the function's documentation
    //~| ERROR use of mutable static is unsafe and requires unsafe block
    //~| NOTE use of mutable static
    //~| NOTE for more information, see issue #71668
    //~| NOTE mutable statics can be mutated by multiple threads
}

macro_rules! unsafe_macro_unsf { () => (unsf()) }
//~^ ERROR call to unsafe function `unsf` is unsafe
//~| NOTE call to unsafe function
//~| NOTE for more information, see issue #71668
//~| NOTE consult the function's documentation
//~| ERROR call to unsafe function `unsf` is unsafe
//~| NOTE call to unsafe function
//~| NOTE for more information, see issue #71668
//~| NOTE consult the function's documentation

// macro_rules! unsafe_macro_multi {
//     ($x:expr) => {{
//         unsf();
//         unsafe_macro_unsf!();
//         $x + bar(BAZ as *const _)
//     }}
// }

pub unsafe fn unsafe_in_macro() {
    //~^ NOTE an unsafe function restricts its caller, but its body is safe by default
    unsafe_macro_unsf!();
    //~^ NOTE in this expansion
    //~| NOTE in this expansion
    //~| NOTE in this expansion
    unsafe_macro_unsf!();
    //~^ NOTE in this expansion
    //~| NOTE in this expansion
    //~| NOTE in this expansion

    // let _ = unsafe_macro_multi!(unsafe_macro_multi!(BAZ));
}

pub unsafe fn unsafe_in_external_macro() {
    // FIXME: https://github.com/rust-lang/rust/issues/112504
    // FIXME: ~^ NOTE an unsafe function restricts its caller, but its body is safe by default
    external_unsafe_macro::unsafe_macro!();
    external_unsafe_macro::unsafe_macro!();
}

pub unsafe fn incomplete_unsafe_op_in_unsafe_fn() {
    //~^ NOTE an unsafe function restricts its caller, but its body is safe by default
    unsafe { unsf(); }
    unsf();
    //~^ ERROR call to unsafe function `unsf` is unsafe
    //~| NOTE call to unsafe function
    //~| NOTE for more information, see issue #71668
    //~| NOTE consult the function's documentation
    unsafe { let _ = unsf(); }
    let _ = unsf();
    //~^ ERROR call to unsafe function `unsf` is unsafe
    //~| NOTE call to unsafe function
    //~| NOTE for more information, see issue #71668
    //~| NOTE consult the function's documentation
}

fn main() {}
