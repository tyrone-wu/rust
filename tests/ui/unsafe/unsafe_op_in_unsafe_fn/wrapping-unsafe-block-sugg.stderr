error[E0133]: call to unsafe function `unsf` is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:13:5
   |
LL |     unsf();
   |     ^^^^^^ call to unsafe function
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> $DIR/wrapping-unsafe-block-sugg.rs:11:1
   |
LL | pub unsafe fn foo() {
   | ^^^^^^^^^^^^^^^^^^^
note: the lint level is defined here
  --> $DIR/wrapping-unsafe-block-sugg.rs:4:9
   |
LL | #![deny(unsafe_op_in_unsafe_fn)]
   |         ^^^^^^^^^^^^^^^^^^^^^^

error[E0133]: call to unsafe function `unsf` is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:17:5
   |
LL |     unsf();
   |     ^^^^^^ call to unsafe function
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: dereference of raw pointer is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:25:13
   |
LL |     let y = *x;
   |             ^^ dereference of raw pointer
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: raw pointers may be null, dangling or unaligned; they can violate aliasing rules and cause data races: all of these are undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> $DIR/wrapping-unsafe-block-sugg.rs:23:1
   |
LL | pub unsafe fn bar(x: *const i32) -> i32 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0133]: dereference of raw pointer is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:29:9
   |
LL |     y + *x
   |         ^^ dereference of raw pointer
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: raw pointers may be null, dangling or unaligned; they can violate aliasing rules and cause data races: all of these are undefined behavior

error[E0133]: use of mutable static is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:38:13
   |
LL |     let y = BAZ;
   |             ^^^ use of mutable static
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> $DIR/wrapping-unsafe-block-sugg.rs:36:1
   |
LL | pub unsafe fn baz() -> i32 {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0133]: use of mutable static is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:42:17
   |
LL |     let z = y + BAZ;
   |                 ^^^ use of mutable static
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error[E0133]: call to unsafe function `bar` is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:46:9
   |
LL |     z + bar(BAZ as *const _)
   |         ^^^^^^^^^^^^^^^^^^^^ call to unsafe function
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior

error[E0133]: use of mutable static is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:46:13
   |
LL |     z + bar(BAZ as *const _)
   |             ^^^ use of mutable static
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: mutable statics can be mutated by multiple threads: aliasing violations or data races will cause undefined behavior

error[E0133]: call to unsafe function `unsf` is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:56:41
   |
LL | macro_rules! unsafe_macro_unsf { () => (unsf()) }
   |                                         ^^^^^^ call to unsafe function
...
LL |     unsafe_macro_unsf!();
   |     -------------------- in this macro invocation
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> $DIR/wrapping-unsafe-block-sugg.rs:74:1
   |
LL | pub unsafe fn unsafe_in_macro() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: this error originates in the macro `unsafe_macro_unsf` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0133]: call to unsafe function `unsf` is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:56:41
   |
LL | macro_rules! unsafe_macro_unsf { () => (unsf()) }
   |                                         ^^^^^^ call to unsafe function
...
LL |     unsafe_macro_unsf!();
   |     -------------------- in this macro invocation
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior
   = note: this error originates in the macro `unsafe_macro_unsf` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0133]: call to unsafe function `unsf` is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:98:5
   |
LL |     unsf();
   |     ^^^^^^ call to unsafe function
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior
note: an unsafe function restricts its caller, but its body is safe by default
  --> $DIR/wrapping-unsafe-block-sugg.rs:95:1
   |
LL | pub unsafe fn incomplete_unsafe_op_in_unsafe_fn() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0133]: call to unsafe function `unsf` is unsafe and requires unsafe block
  --> $DIR/wrapping-unsafe-block-sugg.rs:104:13
   |
LL |     let _ = unsf();
   |             ^^^^^^ call to unsafe function
   |
   = note: for more information, see issue #71668 <https://github.com/rust-lang/rust/issues/71668>
   = note: consult the function's documentation for information on how to avoid undefined behavior

error: aborting due to 12 previous errors

For more information about this error, try `rustc --explain E0133`.
